File: api/download.py
import json
import yt_dlp
import tempfile
import os
import shutil
import base64

def handler(request):
    if request.method != "POST":
        return {
            "statusCode": 405,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": "Invalid method"})
        }

    try:
        body = json.loads(request.body.decode())
    except:
        body = None

    url = body.get("url") if body else None

    if not url:
        return {
            "statusCode": 400,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": "URL must be provided"})
        }

    temp_dir = tempfile.mkdtemp()
    try:
        ydl_opts = {
            "format": "best[ext=mp4]",
            "outtmpl": os.path.join(temp_dir, "%(title)s.%(ext)s"),
            "quiet": True,
            "no_warnings": True,
        }

        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            info = ydl.extract_info(url, download=True)
            file_path = ydl.prepare_filename(info)
            filename = os.path.basename(file_path)

        # Read the file and base64 encode
        with open(file_path, "rb") as f:
            encoded = base64.b64encode(f.read()).decode()

        return {
            "statusCode": 200,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({
                "filename": filename,
                "file_content": encoded
            })
        }

    except Exception as e:
        return {
            "statusCode": 500,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": f"Server error: {str(e)}"})
        }

    finally:
        shutil.rmtree(temp_dir, ignore_errors=True)

File: api/download.js
import ytdl from "ytdl-core";

export default async function handler(req, res) {
  if (req.method !== "POST") {
    res.setHeader("Allow", "POST");
    return res.status(405).json({ error: "Method not allowed" });
  }

  const { url } = req.body;
  if (!url) {
    return res.status(400).json({ error: "URL must be provided" });
  }

  try {
    const info = await ytdl.getInfo(url);
    const title = info.videoDetails.title.replace(/[^a-z0-9]/gi, "_").toLowerCase();

    res.setHeader("Content-Disposition", `attachment; filename="${title}.mp4"`);
    res.setHeader("Content-Type", "video/mp4");

    ytdl(url, { format: "mp4" }).pipe(res);
  } catch (err) {
    return res.status(500).json({ error: "Server error: " + err.message });
  }
}
